### PR TITLE
feat: add FSRS v6 parameter optimizer in pure Go

### LINKED ISSUE #33: Proposal: Add FSRS Optimizer Support via Rust Bindings
Hi maintainers,

I'd like to propose adding the FSRS parameter optimizer to go-fsrs or a separate package.

It seems like we could use cgo to bind to the official fsrs-rs library via its existing C interface (fsrs-rs-c). I'm happy to work on a PR for this if the approach sounds good.

Looking forward to hearing your thoughts, and thank you for all the great work on this project.

### PR DESCRIPTION
**Pull request by cantalupo555**
## Summary

This PR implements the FSRS parameter optimizer in pure Go, addressing #33.

This is the **first pure-language optimizer implementation outside of Rust**, following the reference implementation from [fsrs-rs](https://github.com/open-spaced-repetition/fsrs-rs).

## Features

- **`ComputeParameters()`** - Main API to compute optimal FSRS v6 parameters from review history
- **Pure Go** - No external dependencies (no gonum, no CGO)
- **Full FSRS v6 support** - All 21 parameters optimized
- **Recency weighting** - Recent reviews weighted more heavily
- **Adam optimizer** - With cosine annealing learning rate schedule

## Implementation Details

| Component | Description |
|-----------|-------------|
| `types.go` | FSRSItem, FSRSReview, FSRS v6 constants |
| `dataset.go` | Training data preparation, recency weights |
| `initialization.go` | Ternary search for initial stability parameters |
| `model.go` | Forward pass, power forgetting curve |
| `loss.go` | BCE loss, L2 regularization |
| `gradient.go` | Numerical gradients (central differences) |
| `adam.go` | Adam optimizer, cosine annealing LR |
| `training.go` | Training loop with progress callback |
| `optimizer.go` | Main `ComputeParameters()` API |

## Testing

- **24 unit tests** covering all components
- Benchmarks for gradient computation and training

```bash
go test ./optimizer/... -v
```

## Usage

```go
import "github.com/open-spaced-repetition/go-fsrs/v3/optimizer"

// Prepare review history
items := []optimizer.FSRSItem{
    {Reviews: []optimizer.FSRSReview{
        {Delta: 0, Rating: 3, State: 0},
        {Delta: 1, Rating: 3, State: 1},
        {Delta: 3, Rating: 3, State: 2},
    }},
    // ... more items
}

// Compute optimal parameters
config := optimizer.DefaultConfig()
params, err := optimizer.ComputeParameters(items, config)
if err != nil {
    log.Fatal(err)
}

// Use with go-fsrs scheduler
p := fsrs.DefaultParam()
p.W = params
scheduler := fsrs.NewFSRS(p)
```

## Performance

Performance optimizations (6.5x speedup, matching fsrs-rs) will be submitted in a separate stacked PR to make review easier.

## Future Improvements

- Analytical gradients for additional speedup
- `OptimalRetention()` function
- `Simulate()` for workload prediction

---

Closes #33

### CODE REVIEW

**Comment by L-M-Sherlock** [comment]
The training of fsrs-rs is 15x fast: https://github.com/open-spaced-repetition/fsrs-rs/pull/411 & https://github.com/open-spaced-repetition/fsrs-rs/pull/419

**Comment by cantalupo555** [comment]
> The training of fsrs-rs is 15x fast: [open-spaced-repetition/fsrs-rs#411](https://github.com/open-spaced-repetition/fsrs-rs/pull/411) & [open-spaced-repetition/fsrs-rs#419](https://github.com/open-spaced-repetition/fsrs-rs/pull/419)

@L-M-Sherlock 
I had already implemented forward-mode AD with dual numbers, but I see now that reverse-mode analytical gradients are the right approach for this. I'll switch to that when I get back to the optimizer.

Right now, I'm focused on the next major release.
Thank you for sharing the news about the fsfs-rs optimizer
